### ISSUE TITLE
New custom component for CSP (parabolic trough collectors)

### ISSUE THREAD

**Issue by FranziPl** [oemof/oemof#603]
Hi everyone,

We think about adding a new custom component to model parabolic trough collector fields. It should take into account the temperature and angle of incidence dependence and should be easy to use to help future users, which don't want to build it on their own.
The current concept is divided into 3 parts:

First, the collector irradiation should be calculated with the pvlib, as all the geometric formuals for the cosine losses are allready there. This is a precalculation, which should not be part o the new component, but I mention it for the sake of completeness. In addition I think about  manipulate the time series of irraciance by a helper to take in account the cleanness of the reflectors and heat up losses in the morning, because it would be difficult to include them into the component.

Second, The COP should be calculated. We would use empiric parameters for the optical efficiency and the thermal losses (like described in Janotte 2013: Dynamic performance evaluation of the HelioTrough® collector demonstration loop - towards a new benchmark in parabolic trough qualification). 
The temperatures, which are needed for the thermal losses (input and output temperature of the collector field and ambient temperature )should be given by the user.
This part is also a precalculation, but I would like to include it into the component. It would be a similar diskussion about alias components #594 , like in Issue #591 , but as I didn't understand this concept completely, I cannot say, if it is a good idea here, or not.

Third, the collector should be represented in the component with an input for electrical energy for the pumps and further thermal losses (e.g. for the pipes).

My questions to you are the following:
Would such componet be interesting for you?
How did you model parabolic trough collectors in oemof so far, or didn't you?
Does the approach make sense at all from your point of view?

THanks for your feedback.

**Comment by jakob-wo** [comment]
I would like to add a [Listof Parabolic Trough Projects by the NREL](https://solarpaces.nrel.gov/by-technology/parabolic-trough). It gives an overview of projects, their field dimensions and typical input and output temperatures. Might be helpful to get an impression for everyone who is not too familar with the technology.

**Comment by jakob-wo** [comment]
I see two questions that we need to aswer:

1) How do we wrap or where do we place the precalculation? As you mentioned already that goes along with the question about alias components. I think a solution we find there should be suitable for the CSP as well.

2) I understand the component to model the collector field solely (without any turbine or generator). So what I finally add to my energy system is a `solph.source` with an heat output, is that right? 

We should bring up a code-example or some mockup code to discuss the possible implementations of the component.

**Comment by uvchik** [comment]
Shall we make a web-conference about "Where to place advanced components?" ?

**Comment by jakob-wo** [comment]
👍 yes.
what about Tuesday (18.06.) next week?

**Comment by fwitte** [comment]
Hi, I would like to join you, too! Also, I would like to look into the detailed preprocessing/creating a preprocessing "software". Tuesday would be fine for me, are there any time preferences?

**Comment by uvchik** [comment]
Tuesday is fine for me, too. What about 10:30?

**Comment by jakob-wo** [comment]
10:30h is fine for me, too. I will try to set up a web-conference and let you know how to join (by web-tool or by telephone) until monday.
FranziPl will probably join us.

**Comment by fwitte** [comment]
Web-tool would be best for me!

**Comment by p-snft** [comment]
I would also participate in the Webco, though I will probably not contribute too much to the component.

**Comment by uvchik** [comment]
For me it is more a discussion about the structure.

DFN-Webcos are nice but the are based on Adobe Flash which does not always work as expected.

Alternatively we could try jitsi?

https://meet.jit.si/oemof

**Comment by jakob-wo** [comment]
Hi everyone,
Initially I planed to use DFNconf but I tried it today and I consider it to be too slow and with bad quality (both audio and video transmission).

The tool @uvchik suggested on the other hand seems to be quite convenient and quick, so I propose to use `https://meet.jit.si/oemof`` for our meeting tomorrow, **Tuesday the 18th of June at 10:30h**.
Thanks for coming up with jitsi!

**Comment by jakob-wo** [comment]
Outcome of todays Web-Conf:

- we will set up a new module/repo within the oemof-cosmos, probably called "oemof_heat".
- that module is to hold heat components like the CSP or the Heat Pump (#591) and precalculations associated with or used in the heat components.
- the word "heat components" refers to facade-likes (or alias-like) structures that might be coupled with complex precalculations, like the COP-precalculation in this CSP-example. These components might be placed in a package called `oemof_heat.components`
- the precalculations might be placed in a package called `oemof_heat.cop` (in case of COP precalculations), wich can hold a variety of cop-calculations for the user to choose from. This might include COP-precalculations with TESPy.
- so far it seems that we can realize our heat componets by using the solph.transformer with a cop-precalculation. Increasing unser-friendliness will be done by renaming arguments and adding new arguments.
- if there is the need for a new component with additional constraints, this new component shall be added to `oemof.solph.components` or `oemof.solph.custom` in a generic manner. 

To everyone who attended: please, correct me if I got something wrong or if I missed some point.
To everyone else: please, ask for clearification and leave your comment on our ideas and plans.

**Comment by FranziPl** [comment]
Point 3) just a clearification:
the precalculations should not be included in the components, but in another package (point 4). In the components, they just should be called.
Point 4) 
Maybe the package should have a name like oemof_heat.precalculations with a structure within instead of different packages for the different precalculations. But i think this structure is a detail.

**Comment by FranziPl** [comment]
As discussed in the webcon, here is a block diagram of the text I wrote above:
![concept_github](https://user-images.githubusercontent.com/35954135/59763618-4282d480-929a-11e9-8119-63a1905ec65b.png)

For the precalculation of the cleanliness the formula (1) should be used (derived from [1]), with X as a parameter for clenliness.  For the preheating the calculation has to be defined.
For the efficiency the formulas (2) and (3), also derived from [1] are proposed. 
The electrical consumption and the peripheral thermal losses should be considered by a conversion_factors given by the user.

[1] Janotte, N; et al: Dynamic performance evaluation of the HelioTrough® collector demon-stration loop - towards a new benchmark in parabolic trough qualification, SolarPACES 2013

**Comment by jnnr** [comment]
Sorry that I could not attend the webcon! It seems that you had a productive discussion. jakob-wo updated me. I understand that you discussed two things: The implementation of the CSP specifically, and the structure of new components, in this case heat components. More generally, this relates to the issue So that's my view on the more general question of where to place the code for new (heat) components:.

* I support the approach of separating different things into different bits. Having distinct packages reduces dependencies. 
* Therefore, it seems a good choice to keep any precalculations out of oemof.solph-components. A new solph-component is justified if there is a new constraint to be implemented.. require precalculations
* Specific components like heat pumps, CSP, etc. can be implemented via a combination of precalculations and Transformers/Sources. Therefore, I like the idea of a new repository that helps performing these precalculations. This is parallel to feedinlib or demandlib, which are similar tools that are already widely used.
* I am kind of sceptic about the idea of including facades for oemof.solph-components in the same repo (in this case, oemof_heat). The facades are wrappers for oemof.solph classes, which in my view suggests to keep them close to oemof.solph.
* This would be more transparent. Also, it would be be parallel to feedinlib or demandlib, which are packages that can be used standalone without using oemof.solph. The same could apply to the new package for heat components.

Maybe the latter points should again be discussed in the separate issue [594](https://github.com/oemof/oemof/issues/594)?

Thanks and sorry again for not being able to join the webcon directly!

**Comment by jnnr** [comment]
To move on with this issue, a short wrapup:

We decided to develop heat components in a new repo, called oemof-heat. We can make heat part of the oemof namespace by employing a directory structure /src/oemof/heat, as has been done with oemof-tabular. I created the structure locally and could push it to oemof given the permission to do so.

We still did not decide where facades for CSP, heat pump, etc. should be placed. There are three options:
1. In oemof.heat,
2. in oemof.tabular,
3. in oemof.solph.

The question about facades is discussed here: https://github.com/oemof/oemof/issues/594
